bpf/tc_http_tp.c

```diff
@@ -0,0 +1,680 @@
+        return 0;
+    }
+
+    bpf_skb_pull_data(ctx, ctx->len);
+
+    payload = tcp_payload(ctx);
+
+    if (!payload) {
+        return 0;
+    }
+
+    const unsigned char *end = ctx_data_end(ctx);
+
+    unsigned char *src = payload + nl_offset + 1;
+    unsigned char *dest = src + EXTEND_SIZE;
+
+    if (dest + copy_size > end) {
+        return 0;
+    }
+
+    move_data(dest, src, end, copy_size);
+
+    make_tp_string_skb(src, tp, end);
+
+    payload = tcp_payload(ctx);
+
+    if ((void *)payload > ctx_data_end(ctx)) {
+        return 0;
+    }
+
+    update_ip_csum(ctx);
+
+    struct iphdr *iph = ip_header(ctx);
+
+    if (!iph) {
+        return 0;
+    }
+
+    iph->tot_len = bpf_htons(bpf_ntohs(iph->tot_len) + EXTEND_SIZE);
+
+    return update_tcp_csum(ctx);
+}
+
+static __always_inline void update_tcp_seq(struct __sk_buff *ctx, __u32 extra_bytes) {
+    if (extra_bytes == 0) {
+        return;
+    }
+
+    struct tcphdr *tcp = tcp_header(ctx);
+
+    if (!tcp) {
+        return;
+    }
+
+    __u32 seq = bpf_ntohl(tcp->seq);
+    seq += extra_bytes;
+
+    tcp->seq = bpf_htonl(seq);
+}
+
+static __always_inline void
+get_extra_xmited_bytes(__u32 key, __u32 *extra_bytes, struct tc_http_ctx **http_ctx) {
+    struct tc_http_ctx *ctx = bpf_map_lookup_elem(&tc_http_ctx_map, &key);
+
+    if (ctx) {
+        *extra_bytes = ctx->xtra_bytes;
+        *http_ctx = ctx;
+    } else {
+        *extra_bytes = 0;
+        *http_ctx = NULL;
+    }
+}
+
+static __always_inline void delete_http_ctx(__u32 key) {
+    bpf_map_delete_elem(&tc_http_ctx_map, &key);
+}
+
+static __always_inline void
+update_conn_state_egress(struct tcphdr *tcp, struct tc_http_ctx *http_ctx, __u32 key) {
+    if (tcp->fin) {
+        if (http_ctx->state == ESTABLISHED) {
+            // we've initiated a connection shutdown
+            http_ctx->state = FIN_WAIT_1;
+        } else if (http_ctx->state == CLOSE_WAIT) {
+            // the peer has previously initiated a connection shutdown, we are
+            // now sending our FIN and will wait for the last ACK from the
+            // peer
+            http_ctx->state = LAST_ACK;
+        }
+    }
+}
+
+static __always_inline void
+update_conn_state_ingress(struct tcphdr *tcp, struct tc_http_ctx *http_ctx, __u32 key) {
+    if (tcp->rst) {
+        delete_http_ctx(key);
+        return;
+    }
+
+    // we have previously initiated a connection shutdown
+    if (http_ctx->state == FIN_WAIT_1) {
+        if (tcp->fin && tcp->ack) {
+            // we've entered TIME_WAIT - connection is closed
+            delete_http_ctx(key);
+        } else if (tcp->fin) {
+            // we've only received FIN, but not ACK yet, wait for it
+            http_ctx->state = CLOSING;
+        } else if (tcp->ack) {
+            // we've only received ACK, but no FIN, wait for it
+            http_ctx->state = FIN_WAIT_2;
+        }
+    } else if (http_ctx->state == CLOSING && tcp->ack) {
```

Comment:
You can turn these into OR statements to save some instructions, that way delete_http_ctx will get inlined once only.

Reply:
Excellent idea, will also do it on a follow up PR